Cython/Utility/FunctionArguments.c

```diff
@@ -250,60 +278,427 @@ static int __Pyx_ParseOptionalKeywords(PyObject *kwds, PyObject *const *kwvalues
+                        "%s() got an unexpected keyword argument '%U'",
+                        function_name, key);
+                }
+                #if CYTHON_AVOID_BORROWED_REFS
+                Py_DECREF(key);
+                #endif
+
+                break;
+            }
+            #if CYTHON_AVOID_BORROWED_REFS
+            Py_DECREF(key);
+            #endif
+        }
+    }
+
+    assert(PyErr_Occurred());
+}
+
+static int __Pyx_ParseKeywordDict(
+    PyObject *kwds,
+    PyObject ** const argnames[],
+    PyObject *values[],
+    Py_ssize_t num_pos_args,
+    const char* function_name,
+    int ignore_unknown_kwargs)
+{
+    PyObject** const *name;
+    PyObject** const *first_kw_arg = argnames + num_pos_args;
+    Py_ssize_t kwcount, extracted = 0;
+    int failed = 0;
+
+    // Check if dict is unicode-keys-only and let Python set the error otherwise.
+    if (unlikely(!PyArg_ValidateKeywordArguments(kwds))) return -1;
+
+    #if CYTHON_COMPILING_IN_CPYTHON && CYTHON_ASSUME_SAFE_SIZE
+    kwcount = PyDict_GET_SIZE(kwds);
+    #else
+    kwcount = PyDict_Size(kwds);
+    if (unlikely(kwcount < 0)) return -1;
+    #endif
+
+    // Extract declared keyword arguments.
+    __Pyx_BEGIN_CRITICAL_SECTION(kwds);
+    name = first_kw_arg;
+    while (*name && kwcount > extracted) {
+        PyObject * key = **name;
+        PyObject *value;
+        int found;
+
+        #if __PYX_LIMITED_VERSION_HEX >= 0x030d0000
+        found = PyDict_GetItemRef(kwds, key, &value);
+        #else
+        // There isn't a good way to make this thread-safe before Py3.13.
+        // PyObject_GetItem() would generate lots of expensive KeyErrors.
+        // PyDict_Next() returns borrowed references.
+        value = PyDict_GetItemWithError(kwds, key);
+        if (value) {
+            Py_INCREF(value);
+        } else {
+            if (unlikely(PyErr_Occurred())) {
+                failed = 1;
+                break;
+            };
+        }
+        found = value != NULL;
+        #endif
+
+        if (found) {
+            if (unlikely(found < 0)) {
+                failed = 1;
+                break;
+            };
+            values[name-argnames] = value;
+            extracted++;
+        }
+
+        #if CYTHON_COMPILING_IN_CPYTHON && CYTHON_ASSUME_SAFE_SIZE
+        kwcount = PyDict_GET_SIZE(kwds);
+        #else
+        kwcount = PyDict_Size(kwds);
+        if (unlikely(kwcount < 0)) {
+            failed = 1;
+            break;
+        };
+        #endif
+
+        name++;
+    }
+
+    if (!failed && kwcount > extracted) {
+        if (ignore_unknown_kwargs) {
+            // Make sure the remaining kwargs are not duplicate posargs.
+            failed = __Pyx_ValidateDuplicatePosArgs(kwds, argnames, first_kw_arg, function_name) == -1;
+        } else {
+            // Any remaining kwarg is an error.
+            __Pyx_RejectUnknownKeyword(kwds, argnames, first_kw_arg, function_name);
```

Comment:
I'd move the critical section in here since that does a `PyDict_Next` loop so could _potentially_ go wrong and say the code outside is thread-safe enough that it won't crash and I think that's all that's required here.